feat: implement discussion centers identification feature

JorikSchellekens opened the pull request:
## Overview

This PR implements a comprehensive discussion centers identification feature that analyzes governance proposals to identify key discussion topics and community engagement patterns.

## Key Features

### Backend Changes
- **Community Analysis Service**: Implements cluster-based discussion identification using semantic analysis
- **Generic Cache System**: Adds efficient caching for analysis results with configurable TTL
- **Database Schema**: New migrations for community analysis and generic cache storage
- **Enhanced Agent Service**: Integrates discussion centers analysis into existing agent workflows
- **Deep Research Service**: Provides enhanced proposal analysis capabilities

### Frontend Changes
- **Community Analysis UI**: Interactive visualization components for discussion centers
- **Protocol Management**: Enhanced protocol and space management functionality
- **Cache Integration**: Client-side caching for improved performance
- **Related Proposals**: Service for finding and displaying related governance proposals

### Technical Improvements
- **Database Migrations**: Added 002_community_analysis.sql and 003_generic_cache.sql
- **Repository Layer**: New repositories for community and cache management
- **API Endpoints**: Comprehensive endpoints for community insights and analysis
- **Type Safety**: Enhanced TypeScript types for deep research and community analysis

## Testing
- All Rust code compiles successfully
- UI builds without errors
- Cargo fmt and clippy checks passed
- All existing functionality preserved

## Changes Summary
- 48 files changed
- 3,102 insertions
- 673 deletions
- New services: community analysis, cache management, deep research
- Enhanced UI components with interactive visualizations
- Improved API structure with comprehensive analysis endpoints

JorikSchellekens committed:
feat: add Exa search for related proposals and clickable proposal titles

- Add Exa API integration for semantic search of related proposals
- Create /related-proposals endpoint with authentication
- Add RelatedProposals component displaying search results
- Make proposal titles clickable links to original Snapshot proposals
- Update configuration for optional Exa API key
- Document API integrations and rate limiting in README
- Support graceful degradation when Exa API key not configured

JorikSchellekens committed:
fix: remove needless borrow in Exa service

- Fix clippy warning about needless_borrows_for_generic_args

JorikSchellekens committed:
Add fmt and clippy rules to create-pr

JorikSchellekens committed:
feat: implement discussion centers identification feature

- Add community analysis service with cluster-based discussion identification
- Implement generic cache system for analysis results storage
- Add community analysis endpoints and database schema
- Integrate discussion centers analysis in agent service
- Add community analysis UI components with interactive visualization
- Implement deep research service for enhanced proposal analysis
- Add cache service for efficient data management
- Update protocol and space management functionality
- Add comprehensive analysis endpoints for community insights

yehia67 commented on crates/agent/src/main.rs:
```diff
@@ -47,9 +49,13 @@ async fn main() -> agent::Result<()> {
     };
 
-    let agent_service = AgentService::new(db, config.clone());
+    let agent_service = AgentService::new(db.clone(), config.clone());
```
To avoid cloning db repeatedly
```suggestion
let agent_service = AgentService::new(Arc::clone(&db), config.clone());
```

JorikSchellekens replied:
I've made the cache non optional

yehia67 commented on crates/agent/src/main.rs:
```diff
@@ -47,9 +49,13 @@ async fn main() -> agent::Result<()> {
 
-    let app = create_router(&config, agent_service);
+    // Initialize cache service
+    let cache_repo = CacheRepository::new(db);
+    let cache_service = CacheService::new(cache_repo, None);
+
+    let app = create_router(&config, agent_service, cache_service);
```
Pass optional cache service instead of `None` boilerplate
```suggestion
  let cache_repo = CacheRepository::new(db);
  let cache_service = CacheService::new(cache_repo);
  let app = create_router(&config, agent_service, Some(cache_service));

```

JorikSchellekens replied:
non-optional

yehia67 commented on crates/agent/src/api/routes.rs:
```diff
@@ -40,7 +43,11 @@ impl FromRef<AppState> for Config {
 
 /// Create the API router
-pub fn create_router(config: &Config, agent_service: AgentService) -> Router {
+pub fn create_router(
+    config: &Config,
+    agent_service: AgentService,
+    cache_service: CacheService,
+) -> Router {
```
```suggestion
pub fn create_router(
    config: &Config,
    agent_service: AgentService,
    cache_service: Option<CacheService>,
) -> Router {
```

yehia67 commented on crates/agent/src/db/repositories/mod.rs:
```diff
@@ -1,15 +1,23 @@
 pub use analysis::AnalysisRepository;
 #[allow(unused_imports)]
+pub use cache::CacheRepository;
+#[allow(unused_imports)]
+pub use community::CommunityRepository;
+#[allow(unused_imports)]
```
```suggestion
#[cfg(feature = "cache")]
pub use cache::CacheRepository;

#[cfg(feature = "community")]
pub use community::CommunityRepository;
```

JorikSchellekens replied:
Why make these optional? It's causing build errors to do so

yehia67 replied:
Just to avoid dead code because we import the whole binary
It is failing because we didn't include features in cargo.toml (It is my bad I forget to include on the comment)

```
# inside Cargo.toml
[features]
default = []
cache = []
community = []
```

varex83 replied:
We can leave them as a dead code, making them feature-based is not a good way, since we don't use agent as a library

yehia67 replied:
My bad,  I misunderstood how #[cfg(feature = ...)] works. It requires proper feature flags in Cargo.toml. Please ignore my earlier suggestion, I’ll mark this as resolved

JorikSchellekens committed:
Fix PR review issues and UI lint warnings

- Fix database cloning optimization in main.rs
- Make cache service optional in create_router and update handlers
- Remove unused conditional compilation attributes from repositories
- Fix cache service constructor parameters
- Remove unused navigateToPage parameter from ProposalCard components
- Update all cache-related handlers to handle optional cache service
- Apply cargo fmt and clippy fixes
- Resolve ESLint warnings in UI components

JorikSchellekens committed:
Update crates/agent/src/main.rs

Co-authored-by: Yehia Tarek <yehia@nethermind.io>

JorikSchellekens committed:
Refactor cache service to eliminate boilerplate

- Simplify CacheService::new to not require optional config parameter
- Make cache service non-optional in create_router and AppState
- Remove if let Some(cache_service) boilerplate from handlers
- Pass cache service directly without Some() wrapper in main.rs
- Clean up unused imports and improve code maintainability

JorikSchellekens committed:
Merge develop into feat_identify_discussion_centers

- Resolved merge conflicts in handlers.rs, routes.rs, models/mod.rs, and README.md
- Integrated Swagger UI documentation from develop branch
- Merged API documentation and API integrations sections in README
- Updated imports to use models from centralized location
- Maintained cache service refactoring improvements

JorikSchellekens committed:
Fix UI merge conflicts and ensure successful compilation

- Resolved merge conflicts in analyzer-client.tsx by combining useMemo optimization with domain field
- Fixed merge conflicts in proposal-page.tsx by merging useCallback optimization with forceRefresh parameter
- Removed undefined variables and simplified handleBackClick function
- UI now compiles successfully with optimized production build
- All TypeScript errors resolved
